File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        
        vector<vector<int>> ans;
        
        queue<TreeNode*> q;
        
        if(!root) return ans;
        
        bool ltr = true;
        
        q.push(root);
        
        while(!q.empty()){
            
            int n = q.size();
            vector<int> row(n);
            
            for(int i=0;i<n;i++){
                
                TreeNode* curr = q.front();
                q.pop();
                
                int index = ltr ? i : (n-i-1);
                row[index] = curr->val;
                
                if(curr->left) q.push(curr->left);
                if(curr->right) q.push(curr->right);
            }
            
            ltr = !ltr;
            ans.push_back(row);
            
        }
                
        return ans;
        
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        ans = []
        
        q = []
        
        ltr = True
        
        if(root):
            q.append(root)
            
        while(len(q)!=0):
            
            n = len(q)
            temp = [0]*n
            
            for i in range(n):
                
                curr = q[0]
                q.pop(0)
                
                idx = i if ltr==True else n-i-1
                
                temp[idx] = curr.val
                
                if curr.left:
                    q.append(curr.left)
                
                if curr.right:
                    q.append(curr.right)
                
            ltr = False if ltr is True else True
            ans.append(temp)
        
        return ans
                
        